hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateTopLevelDagVisitor.java

```diff
@@ -496,7 +507,9 @@ public Vertex onHashJoin(JoinHashPhysicalRel rel) {
 
     @Override
     public Vertex onStreamToStreamJoin(StreamToStreamJoinPhysicalRel rel) {
-        JetJoinInfo joinInfo = rel.joinInfo(parameterMetadata);
+        dagBuildContext.setTable(null);
```

Comment:
this method does not use `dagBuildContext` (there are more methods like that). It is good to set it for consistency, but maybe instead of having `dagBuildContext` as a field in `CreateTopLevelDagVisitor` we could create context only when needed? Then there would be no risk of forgetting to update it.

Reply:
One mutable or many immutable instances - seems about the same to me, and it's invisible from the user's point of view. I have a gut feeling that we might need to add more and more context info to it, and mutable makes sense.

But it's true, in this case we don't have to bother setting the context objects...

Reply:
mutable object is much harder to reason about. I strongly prefer immutable, if possible.

In nested loops PR I needed to have some non-locality when visiting nodes. Eg. index scan or full scan is interpreted differently if it is on right side of `JoinNestedLoopPhysicalRel` or not. My idea for that was to have different visitor that captures such state, so there can be less `if`s.

Reply:
>mutable object is much harder to reason about.
Can you be specific here? From user's POV, it's immutable. It's also single-threaded (no races), and the lifecycle is simple: before each call to user method, set the context (except for a few cases where we use the context ourselves). In this specific case I prefer mutable context, as it produces less litter and it's easier to implement, though these advantages are minor.

Regarding the nested loops PR note, does the mutable context somehow interfere with it?

Reply:
Mutable context should not interfere with nested loops PR. Nested loops probably needs a stack to support nested nested loops (not implemented yet), not simple plain context. The stack might be created of nested visitors.